### PR TITLE
BUG: timedelta64(NaT) incorrectly treated as datetime in some dataframe ops

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/ops/__init__.py [hunk 1]
```diff
@@ -498,8 +498,17 @@ def column_op(a, b):
         # in which case we specifically want to operate row-by-row
         assert right.index.equals(left.columns)
 
-        def column_op(a, b):
-            return {i: func(a.iloc[:, i], b.iloc[i]) for i in range(len(a.columns))}
+        if right.dtype == "timedelta64[ns]":
+            # ensure we treat NaT values as the correct dtype
+            right = np.asarray(right)
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: what if you *always* do ``right = np.asarray(right)``? (and just use a positional indexer in ``column_op``)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops/__init__.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: That will be incorrect for EA dtypes

Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: (well, not incorrect, but potentially costly)

Comment (TomAugspurger) [reply pandas/core/ops/__init__.py]: Given that Jeff and I had the same question, I think that a comment here would be helpful.

Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: i like this heuristic, will do

**Review by TomAugspurger — APPROVED** [review]
LGTM. Could a user hit this from normal code? (i.e. do we need a release note?)

**Comment by jbrockmendel** [comment]
> Could a user hit this from normal code? (i.e. do we need a release note?)

Yes.  e.g. the test this implements incorrectly raises TypeError on master.  Will update with release note.

**Comment by jbrockmendel** [comment]
@jreback gentle ping

File: doc/source/whatsnew/v0.25.2.rst [hunk 2]
```diff
@@ -18,7 +18,7 @@ Categorical
 
 Datetimelike
 ^^^^^^^^^^^^
-
+- Bug in :class:`DataFrame` arithmetic operations when operating with a :class:`Series` with dtype `'timedelta64[ns]'` (:issue:`28049`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.2.rst]: can move to 1.0

**Review by jreback — APPROVED** [review]
lgtm, just move the note to 1.0

File: doc/source/whatsnew/v0.25.2.rst (hunk 2 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v0.25.2.rst]: updated+green

File: pandas/core/ops/__init__.py [hunk 3]
```diff
@@ -499,8 +499,19 @@ def column_op(a, b):
...
-        def column_op(a, b):
-            return {i: func(a.iloc[:, i], b.iloc[i]) for i in range(len(a.columns))}
+        if right.dtype == "timedelta64[ns]":
+            # ensure we treat NaT values as the correct dtype
+            # Note: we do not do this unconditionally as it may be lossy or
+            #  expensive for EA dtypes.
+            right = np.asarray(right)
+
+            def column_op(a, b):
+                return {i: func(a.iloc[:, i], b[i]) for i in range(len(a.columns))}
```
Comment (WillAyd) [inline pandas/core/ops/__init__.py]: Just to double check - is the second argument to `func` supposed to be accessed via `.iloc` here as well?

Side note - an alternate approach for by column iteration is to call `df.columns`; not sure if there is a perf difference but have seen both in code base

**Review by WillAyd — CHANGES_REQUESTED** [review]
Minor clarification point. Was about to merge but want to double check before.

Note I fixed up unneeded changes to the 0.25.2 whatsnew so would need to pull locally if a change required here (or edit directly in GH)

File: pandas/core/ops/__init__.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: > is the second argument to func supposed to be accessed via .iloc here as well?

No, at this point `b` is an ndarray.

> Side note - an alternate approach for by column iteration is to call df.columns; not sure if there is a perf difference but have seen both in code base

That runs in to difficulties if there are duplicate columns.

Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: > not sure if there is a perf difference

Might be worth looking at using `iat` instead of `iloc` for perf

**Comment by WillAyd** [comment]
Sounds good - merge away

**Comment by jreback** [comment]
lgtm merge on green

**Comment by jbrockmendel** [comment]
@WillAyd gentle ping

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @jbrockmendel

**Pull request merged**